demo/lightning/checkpoint/simulated/llama2.py

```diff
@@ -0,0 +1,289 @@
+        super().__init__(**kwargs)
+        if use_fsspec:
+            self.reader = FF.FsspecReader(path)
+            self.writer = FF.FsspecWriter(path, sync_files=False)
+        else:
+            self.writer = GCSDistributedWriter(path, project, None)
+            self.reader = GCSDistributedReader(path, project, None)
+
+    def save_checkpoint(self,
+                        checkpoint: Dict[str, torch.Tensor],
+                        filepath: str,
+                        storage_options: Optional[Dict] = None) -> None:
+        """
+        Saves the model's state dictionary to a specified file path in GCS.
+        torch.distributed.checkpoint.save contains the core logic for saving
+        model shards.
+        Source code for FSDP.save_checkpoint can be found at
+        https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/strategies/fsdp.py#L553 .
+        Args:
+            checkpoint (Dict[str, torch.Tensor]): The model's state dictionary
+            containing tensor weights.
+            filepath (str): The path where the checkpoint will be saved.
+            storage_options (Optional[Dict]): Additional storage options
+            (if any).
+
+        This method uses the GCS writer to save the checkpoint.
+        """
+        dist_cp.save(state_dict=checkpoint,
+                     checkpoint_id=filepath,
+                     storage_writer=self.writer)
+
+    def load_checkpoint(self, checkpoint_path: str,
+                        initial_state_dict: Dict) -> None:
+        """
+        Loads a model's state dictionary from a specified checkpoint file in
+        GCS.
+        torch.distributed.checkpoint.load contains the core logic of loading
+        sharded model weights.
+        Source code for FSDP.load_checkpoint can be found at
+        https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/strategies/fsdp.py#L589 .
+
+        For torch.distributed.checkpoint.load to work properly the
+        template_state_dict should have model format.
+        The values of the keys will be overwritten with new values from
+        sharded checkpoint after load is successful.
+
+        Args:
+            checkpoint_path (str): The path to the checkpoint file to be
+            loaded.
+
+        This method reads the checkpoint from GCS and updates the model's
+        state dictionary.
+        """
+        dist_cp.load(state_dict=initial_state_dict,
+                     checkpoint_id=checkpoint_path,
+                     storage_reader=self.reader)
+
+
+def setup(rank: int, world_size: int) -> None:
+    """Sets up the distributed environment."""
+    os.environ['MASTER_ADDR'] = MASTER_ADDR
+    os.environ['MASTER_PORT'] = MASTER_PORT
+    dist.init_process_group("gloo", rank=rank, world_size=world_size)
+
+
+def cleanup() -> None:
+    dist.destroy_process_group()
+
+
+def time_checkpoint_operation(benchmark_strategy: BenchmarkStrategy,
```

Comment:
Out of scope for this CL, but in the future we can save a lot of duplicate code if we use a timer decorator function defined in a single location and used on each method with `@timer_function`

Reply:
Ack